### PR TITLE
Refresh Windows Domainless gMSA plugin credentials

### PR DESCRIPTION
**Pull request by arun-annamalai**
### Summary

The domainless gMSA plugin requires taskExecutionRole IAM credentials to access customer resources. These credentials are populated once at the beginning of task start, but they need to be updated each time the credentials are vended down to the agent. 

### Implementation details

This change is primarily implemented in the refresh handler. Everytime an execution role gets vended down, the refresh handler should populate the Windows registry key where the plugin credentials are stored. 

### Testing

This change was unit tested

I have also let a task run for 2 days and then checked out the instance logs and I do see that the credentials are being set correctly after a refresh from ACS (have logs I can send) 

New tests cover the changes: 
yes
### Description for the changelog

Refresh Windows Domainless gMSA plugin credentials
### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/acs/handler/refresh_credentials_handler_test.go [hunk 1]
```diff
@@ -233,6 +236,67 @@ func TestCredentialsMessageNotAckedWhenTaskNotFound(t *testing.T) {
 	cancel()
 }
 
+// TestCredentialsMessageNotAckedWhenDomainlessGMSACredentialsNotSet tests if credential messages
+// are not acked when the task arn in the message is not found in the task
+// engine
+func TestCredentialsMessageNotAckedWhenDomainlessGMSACredentialsError(t *testing.T) {
+	testCases := []struct {
```
Comment (mythri-garaga) [inline agent/acs/handler/refresh_credentials_handler_test.go:243]: question(non-blocking): Since the test name suggests that we are testing a very specific scenario, do we need table-driven tests here?

Comment (arun-annamalai) [reply agent/acs/handler/refresh_credentials_handler_test.go:243]: I have tested this manually over the last 2 days (have logs on the instance I can send)
You are right that we don't need table tests, but it was just in case for future extensibility. I can remove if needed.

File: agent/acs/handler/refresh_credentials_handler_test.go [hunk 2]
```diff
@@ -233,6 +236,67 @@ func TestCredentialsMessageNotAckedWhenTaskNotFound(t *testing.T) {
 	cancel()
 }
 
+// TestCredentialsMessageNotAckedWhenDomainlessGMSACredentialsNotSet tests if credential messages
+// are not acked when the task arn in the message is not found in the task
```
Comment (amogh09) [inline agent/acs/handler/refresh_credentials_handler_test.go:240]: Is this test testing the case when "the task arn in the message is not found in the task"? I think the test is for the case when setting domainless GMSA creds failed?

**Review by mythri-garaga — APPROVED** [review]

File: agent/acs/handler/refresh_credentials_handler.go [hunk 3]
```diff
@@ -160,6 +165,15 @@ func (refreshHandler *refreshCredentialsHandler) handleSingleMessage(message *ec
 		}
 		if roleType == credentials.ExecutionRoleType {
 			task.SetExecutionRoleCredentialsID(aws.StringValue(message.RoleCredentials.CredentialsId))
+			// Refresh domainless gMSA plugin credentials
+			if task.RequiresDomainlessCredentialSpecResource() {
```
Comment (amogh09) [inline agent/acs/handler/refresh_credentials_handler.go:169]: How about we perform this check in the setter function (and maybe rename it)? That way the refresh handler would be simpler and have one less responsibility.

Also I think `TestHandleRefreshMessageAckedWhenCredentialsUpdatedDomainlessGMSA` would become much simpler as we will be able to split it into two tests - 
1. Acks are sent when domainless gMSA setter does not fail (we might even be able to use an existing test for this)
2. Setter function for windows does not attempt to update registry if task does not require domainless GMSA.

File: agent/acs/handler/refresh_credentials_handler_test.go (hunk 2 above)
Comment (arun-annamalai) [reply agent/acs/handler/refresh_credentials_handler_test.go:240]: ah yeah too much copy paste. Let me fix this comment, you are correct that this test case is for when setting domainless creds fails.

File: agent/acs/handler/refresh_credentials_handler.go [hunk 4]
```diff
@@ -160,6 +165,15 @@ func (refreshHandler *refreshCredentialsHandler) handleSingleMessage(message *ec
...
+				err = setDomainlessGMSATaskExecutionRoleCredentialsImpl(iamRoleCredentials, task.Arn)
+				if err != nil {
+					seelog.Errorf("Unable to SetDomainlessGMSATaskExecutionRoleCredentials for task %s, err: %v messageId: %s", taskArn, err, messageId)
+					return fmt.Errorf("unable to SetDomainlessGMSATaskExecutionRoleCredentials %v", err)
```
[the remark this answers is not in the document]
Comment (arun-annamalai) [reply agent/acs/handler/refresh_credentials_handler.go]: Done

File: agent/acs/handler/refresh_credentials_handler.go (hunk 3 above)
Comment (arun-annamalai) [reply agent/acs/handler/refresh_credentials_handler.go:171]: Addressed both. 

Item 1 is in Function TestHandleRefreshMessageAckedWhenCredentialsUpdated


Item 2 is in Function TestHandleRefreshMessageAckedAndNoSetDomainlessCredentialsWhenNonDomainlessGMSATask

File: agent/acs/handler/refresh_credentials_handler.go [hunk 5]
```diff
@@ -160,6 +167,15 @@ func (refreshHandler *refreshCredentialsHandler) handleSingleMessage(message *ec
 		}
 		if roleType == credentials.ExecutionRoleType {
 			task.SetExecutionRoleCredentialsID(aws.StringValue(message.RoleCredentials.CredentialsId))
+			// Refresh domainless gMSA plugin credentials
+			if task.RequiresDomainlessCredentialSpecResource() {
+				err = checkAndSetDomainlessGMSATaskExecutionRoleCredentialsImpl(iamRoleCredentials, task.Arn, task.RequiresDomainlessCredentialSpecResource())
```
Comment (amogh09) [inline agent/acs/handler/refresh_credentials_handler.go:172]: I think we should pass `task` to this function and let it extract the ARN and figure out if it's a domainless gMSA task. Passing the result of `task.RequiresDomainlessCredentialSpecResource()` still leaks implementation details into the ACS refresh handler.

File: agent/acs/handler/refresh_credentials_handler_linux.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+package handler
+
+import (
+	"github.com/aws/amazon-ecs-agent/ecs-agent/credentials"
+)
+
+func checkAndSetDomainlessGMSATaskExecutionRoleCredentials(iamRoleCredentials credentials.IAMRoleCredentials, taskArn string, domainlessGMSATask bool) error {
+	// exit early if the task does not need domainless gMSA
+	if !domainlessGMSATask {
```
Comment (amogh09) [inline agent/acs/handler/refresh_credentials_handler_linux.go:25]: Why do we need this if? Isn't it redundant?

File: agent/acs/handler/refresh_credentials_handler_test.go [hunk 7]
```diff
@@ -233,50 +236,243 @@ func TestCredentialsMessageNotAckedWhenTaskNotFound(t *testing.T) {
...
+		})
 	}
+}
 
-	// Wait till we get an ack from the ackBuffer
-	select {
-	case <-ctx.Done():
+// TestHandleRefreshMessageAckedWhenCredentialsUpdated tests that a credential message
+// is ackd when the credentials are updated successfully and the domainless gMSA plugin credentials are updated successfully
+func TestHandleRefreshMessageAckedWhenCredentialsUpdated(t *testing.T) {
```
Comment (amogh09) [inline agent/acs/handler/refresh_credentials_handler_test.go:311]: Do we need this new test? I think `TestHandleRefreshMessageAckedWhenCredentialsUpdated` already covers this case. 

Refresh handler should care only about two cases (once we move `task.RequiresDomainlessCredentialSpecResource()` call to the implementation of checkAndSet function) - 
1. `checkAndSetDomainlessGMSATaskExecutionRoleCredentialsImpl` returns an `error`
2. `checkAndSetDomainlessGMSATaskExecutionRoleCredentialsImpl` returns `nil`

Comment (amogh09) [reply agent/acs/handler/refresh_credentials_handler_test.go:311]: If we want an explicit test then we can add this test with a single case inside (checkAndSet returns `nil`).

File: agent/acs/handler/refresh_credentials_handler_linux.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+package handler
+
+import (
+	"github.com/aws/amazon-ecs-agent/agent/api/task"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/credentials"
+)
+
+func checkAndSetDomainlessGMSATaskExecutionRoleCredentials(iamRoleCredentials credentials.IAMRoleCredentials, task *task.Task) error {
+	// exit early if the task does not need domainless gMSA
+	if !task.RequiresDomainlessCredentialSpecResource() {
```
[the remark this answers is not in the document]
Comment (amogh09) [reply agent/acs/handler/refresh_credentials_handler_linux.go:26]: Spoke offline to Arun. We will need this if block later.

File: agent/api/container/container.go [hunk 9]
```diff
@@ -1366,6 +1366,16 @@ func (c *Container) RequiresCredentialSpec() bool {
 	return true
 }
 
+// RequiresDomainlessCredentialSpec checks if container needs a domainless credentialspec resource
+func (c *Container) RequiresDomainlessCredentialSpec() bool {
+	credSpec, err := c.getCredentialSpec()
```
[the remark this answers is not in the document]
Comment (amogh09) [reply agent/api/container/container.go:1371]: Talked to Arun offline. If we log here then that can get noisy for non-domainless gMSA tasks. So we will skip adding logs here.

File: agent/acs/handler/refresh_credentials_handler_test.go [hunk 10]
```diff
@@ -234,49 +237,158 @@ func TestCredentialsMessageNotAckedWhenTaskNotFound(t *testing.T) {
...
+		domainlessGMSATaskExpectedInput bool
+		containers                      []*apicontainer.Container
+	}{
+		{
+			name:       "EmptyTaskSucceeds",
+			taskArn:    taskArn,
+			containers: []*apicontainer.Container{},
+		},
+		{
+			name:                            "DomainlessTaskSucceeds",
```
[the remark this answers is not in the document]
Comment (arun-annamalai) [reply agent/acs/handler/refresh_credentials_handler_test.go:254]: yes will be deleting 1 test case

File: agent/acs/handler/refresh_credentials_handler_test.go [hunk 11]
```diff
@@ -234,49 +237,158 @@ func TestCredentialsMessageNotAckedWhenTaskNotFound(t *testing.T) {
...
+			checkAndSetDomainlessGMSATaskExecutionRoleCredentialsImpl = func(iamRoleCredentials credentials.IAMRoleCredentials, task *apitask.Task) error {
+				if tc.taskArn != task.Arn {
+					return errors.New(fmt.Sprintf("Expected taskArnInput to be %s, instead got %s", tc.taskArn, task.Arn))
+				}
 
-	// Wait till we get an ack from the ackBuffer
-	select {
-	case <-ctx.Done():
-	}
+				if tc.domainlessGMSATaskExpectedInput != task.RequiresDomainlessCredentialSpecResource() {
```
[the remark this answers is not in the document]
Comment (arun-annamalai) [reply agent/acs/handler/refresh_credentials_handler_test.go:285]: removing this check

File: agent/acs/handler/refresh_credentials_handler.go [hunk 12]
```diff
@@ -160,6 +167,13 @@ func (refreshHandler *refreshCredentialsHandler) handleSingleMessage(message *ec
 		}
 		if roleType == credentials.ExecutionRoleType {
 			task.SetExecutionRoleCredentialsID(aws.StringValue(message.RoleCredentials.CredentialsId))
+			// Refresh domainless gMSA plugin credentials if needed
+			err = checkAndSetDomainlessGMSATaskExecutionRoleCredentialsImpl(iamRoleCredentials, task)
+			if err != nil {
+				seelog.Errorf("Unable to SetDomainlessGMSATaskExecutionRoleCredentials for task %s, err: %v messageId: %s", taskArn, err, messageId)
+				return errors.Wrap(err, "unable to SetDomainlessGMSATaskExecutionRoleCredentials")
+			}
+			seelog.Infof("Successfully SetDomainlessGMSATaskExecutionRoleCredentials for task %s messageId: %s", taskArn, messageId)
```
[the remark this answers is not in the document]
Comment (arun-annamalai) [reply agent/acs/handler/refresh_credentials_handler.go:176]: correct, will be fixing this

[the remark this answers is not in the document]
Comment (arun-annamalai) [reply agent/acs/handler/refresh_credentials_handler.go]: I moved the log line into the function 

credentialspec.SetTaskExecutionCredentialsRegKeys
